stackslib/src/net/api/gettransaction.rs

```diff
@@ -0,0 +1,253 @@
+    HttpResponsePreamble, HttpServerError,
+};
+use crate::net::httpcore::{
+    request, HttpPreambleExtensions, HttpRequestContentsExtensions, RPCRequestHandler,
+    StacksHttpRequest, StacksHttpResponse,
+};
+use crate::net::p2p::PeerNetwork;
+use crate::net::{Error as NetError, StacksNodeState, TipRequest};
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct TransactionResponse {
+    pub index_block_hash: StacksBlockId,
+    pub tx: String,
+}
+
+#[derive(Clone)]
+pub struct RPCGetTransactionRequestHandler {
+    pub txid: Option<Txid>,
+}
+impl RPCGetTransactionRequestHandler {
+    pub fn new() -> Self {
+        Self { txid: None }
+    }
+}
+
+/// Decode the HTTP request
+impl HttpRequest for RPCGetTransactionRequestHandler {
+    fn verb(&self) -> &'static str {
+        "GET"
+    }
+
+    fn path_regex(&self) -> Regex {
+        Regex::new(r#"^/v3/transactions/(?P<txid>[0-9a-f]{64})$"#).unwrap()
+    }
+
+    fn metrics_identifier(&self) -> &str {
+        "/v3/transactions/:txid"
+    }
+
+    /// Try to decode this request.
+    /// There's nothing to load here, so just make sure the request is well-formed.
+    fn try_parse_request(
+        &mut self,
+        preamble: &HttpRequestPreamble,
+        captures: &Captures,
+        query: Option<&str>,
+        _body: &[u8],
+    ) -> Result<HttpRequestContents, Error> {
+        if preamble.get_content_length() != 0 {
+            return Err(Error::DecodeError(
+                "Invalid Http request: expected 0-length body for GetTransaction".to_string(),
+            ));
+        }
+
+        let txid = request::get_txid(captures, "txid")?;
+        self.txid = Some(txid);
+
+        Ok(HttpRequestContents::new().query_string(query))
+    }
+}
+
+impl RPCRequestHandler for RPCGetTransactionRequestHandler {
+    /// Reset internal state
+    fn restart(&mut self) {
+        self.txid = None;
+    }
+
+    /// Make the response
+    fn try_handle_request(
+        &mut self,
+        preamble: HttpRequestPreamble,
+        contents: HttpRequestContents,
+        node: &mut StacksNodeState,
+    ) -> Result<(HttpResponsePreamble, HttpResponseContents), NetError> {
+        if !is_transaction_log_enabled() {
+            return StacksHttpResponse::new_error(
+                &preamble,
+                &HttpNotImplemented::new("Transaction log is not enabled".into()),
+            )
+            .try_into_contents()
+            .map_err(NetError::from);
+        }
+
+        let tip = match node.load_stacks_chain_tip(&preamble, &contents) {
+            Ok(tip) => tip,
+            Err(error_resp) => {
+                return error_resp.try_into_contents().map_err(NetError::from);
+            }
+        };
+
+        let txid = self
+            .txid
+            .take()
+            .ok_or(NetError::SendError("`txid` no set".into()))?;
+
+        node.with_node_state(|_network, _sortdb, chainstate, _mempool, _rpc_args| {
```

Comment:
It seems like you are trying to answer to separate queries here:

* What is the transaction for the given txid?
* Is the transaction on the canonical Stacks fork?

Answering the latter can be very, very expensive with the approach you have taken.  You're effectively scanning an unbounded number of blocks' transactions, and doing MARF reads on each block relative to the chain tip.

I don't know if there is value in only returning transactions if they are on the canonical chain, since nearly all the time they will be.  Instead, I think it's just as effective to return the transaction and the list of block IDs in which the node knows it occurs (which will almost always be one block).  It would then be on the user to verify that those blocks are canonical.

Reply:
I asked in another thread how it could be possible for a transaction to be mined in multiple different Nakamoto blocks, and I think the only way this could happen is if there is a _deep_ Nakamoto fork.  This would only happen if there was a Bitcoin fork, and the second sibling Bitcoin block was mined substantially later than the first sibling.  Nakamoto would preserve both Bitcoin forks and both Stacks forks atop them, and would write a transaction log entry for the same transaction in both forks.

It's not impossible, but it would be so unlikely that it might not be worth solving at this time.

Reply:
i like this approach way more than the current one, it will make the code way easier and extracting the tx from a block given the index_block_hash is straightforward.

I still get dozens of duplicated txid (with different index_block_hash) in my testnet transaction_log, i will try to understand why this is happening